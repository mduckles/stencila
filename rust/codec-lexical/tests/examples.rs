use std::{fs::read_to_string, path::PathBuf};

use codec::{
    common::{eyre::Result, glob::glob},
    format::Format,
    EncodeOptions,
};

use codec_lexical::{decode, encode};
use common_dev::insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};

/// Decode each example of a Lexical document and create JSON and Lexical snapshots
/// including snapshots for losses
#[test]
fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*";

    for path in glob(&format!("{pattern}.lexical"))?
        .chain(glob(&format!("{pattern}.koenig"))?)
        .flatten()
    {
        let format = Format::from_path(&path);
        let ext = format.extension();

        let name = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(&[".", &ext].concat()).map(String::from))
            .expect("should have format suffix");

        let original = read_to_string(path)?;

        let (article, info) = decode(&original, None)?;

        assert_json_snapshot!(format!("{name}.json"), article);
        assert_yaml_snapshot!(format!("{name}.decode.losses"), info.losses);

        let (lexical, info) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format.clone()),
                compact: Some(false),
                ..Default::default()
            }),
        )?;

        assert_snapshot!(format!("{name}.{ext}"), lexical);
        assert_yaml_snapshot!(format!("{name}.encode.losses"), info.losses);

        let (lexical, ..) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format),
                standalone: Some(true),
                compact: Some(false),
                ..Default::default()
            }),
        )?;
        assert_snapshot!(format!("{name}.standalone.{ext}"), lexical);
    }

    Ok(())
}
