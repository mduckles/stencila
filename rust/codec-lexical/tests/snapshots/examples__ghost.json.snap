---
source: rust/codec-lexical/tests/examples.rs
expression: article
---
{
  "type": "Article",
  "content": [
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "A document for testing Stencila's integration with Ghost. This document is used as a test fixture, so be careful making changes because it is likely to affect those tests. "
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Paragraph with "
          }
        },
        {
          "type": "Strong",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "bold"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Emphasis",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "italic"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Emphasis",
          "content": [
            {
              "type": "Strong",
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "bold and italic"
                  }
                }
              ]
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Strikeout",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "strikethrough"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Underline",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "underline"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "CodeInline",
          "code": {
            "string": "code"
          }
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Subscript",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "subscript"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Superscript",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "superscript"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " "
          }
        },
        {
          "type": "Annotation",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "highlight"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " and a "
          }
        },
        {
          "type": "Text",
          "value": {
            "string": "LOST Link"
          }
        }
      ]
    },
    {
      "type": "Heading",
      "level": 1,
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Heading 1"
          }
        }
      ]
    },
    {
      "type": "Heading",
      "level": 2,
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Heading 2"
          }
        }
      ]
    },
    {
      "type": "Heading",
      "level": 3,
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Heading 3"
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": []
    },
    {
      "type": "QuoteBlock",
      "content": [
        {
          "type": "Paragraph",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "A quote"
              }
            }
          ]
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "LOST Unknown (aside)"
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "LOST List"
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "LOST List"
          }
        }
      ]
    },
    {
      "type": "ThematicBreak"
    },
    {
      "type": "ImageObject",
      "contentUrl": "https://stencila.ghost.io/content/images/2025/01/stencila-128.png"
    },
    {
      "type": "CodeBlock",
      "code": {
        "string": "print(\"Hello world\")"
      },
      "programmingLanguage": "Python"
    },
    {
      "type": "Table",
      "rows": [
        {
          "type": "TableRow",
          "cells": [
            {
              "type": "TableCell",
              "content": [
                {
                  "type": "Paragraph",
                  "content": [
                    {
                      "type": "Text",
                      "value": {
                        "string": "A"
                      }
                    }
                  ]
                }
              ]
            },
            {
              "type": "TableCell",
              "content": [
                {
                  "type": "Paragraph",
                  "content": [
                    {
                      "type": "Text",
                      "value": {
                        "string": "B"
                      }
                    }
                  ]
                }
              ]
            }
          ],
          "rowType": "HeaderRow"
        },
        {
          "type": "TableRow",
          "cells": [
            {
              "type": "TableCell",
              "content": [
                {
                  "type": "Paragraph",
                  "content": [
                    {
                      "type": "Text",
                      "value": {
                        "string": "1"
                      }
                    }
                  ]
                }
              ]
            },
            {
              "type": "TableCell",
              "content": [
                {
                  "type": "Paragraph",
                  "content": [
                    {
                      "type": "Text",
                      "value": {
                        "string": "2"
                      }
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "type": "RawBlock",
      "format": "html",
      "content": {
        "string": "<span>HTML card</span>"
      }
    }
  ]
}
